swarm/src/connection.rs

```diff
@@ -186,10 +189,22 @@ where
         max_negotiating_inbound_streams: usize,
         idle_timeout: Duration,
     ) -> Self {
-        let initial_protocols = gather_supported_protocols(&handler);
-        if !initial_protocols.is_empty() {
+        let local_supported_protocols = handler
+            .listen_protocol()
+            .upgrade()
+            .protocol_info()
+            .map(|i| (AsStrHashEq(i), true))
+            .collect::<HashMap<_, _>>();
```

Comment:
Can we go back to just calling a helper function here please? That would remove some of the noise in this function.